### ISSUE TITLE
More Widgets / Widgets for where Select2 is unsuitable / Replacement V2 style 'minimal' widgets

### ISSUE THREAD

**Issue by techdragon** [yourlabs/django-autocomplete-light#581]
There is much to love about v3... However for me (and I suspect others) it comes with a massive drawback. The previous v2 widgets were MUCH less obtrusive than the new V3 Select2 powered one is.

I was using DAL specifically because it was the only autocomplete library that didnt require select2 and worked out of the box with "Material Design Style" text inputs. I spent several days trying to get various autocomplete mechanisms working with my design before trying DAL and discovering that V2 worked out of the box with my site's UI. 

Examples of the sort of design where the new Select2 widget feels unsuitable are:
https://www.google.com/design/spec/components/text-fields.html#text-fields-labels
http://mdbootstrap.com/components/forms/ 
http://fezvrasta.github.io/bootstrap-material-design/#input
https://bootswatch.com/paper/#forms

The V2 widgets (text input & tag) were much more compatible out of the box with these sorts of styles, and it's extremely frustrating that all the Python side improvements in DAL V3 have to come at the expense of more complex CSS & JS required to subdue Select2 to fit what is becoming an increasingly popular and sometimes even required design paradigm.

Can we get a more minimal select2 powered widget that works more like the old widgets? a port of the old widgets to the new API? or something to make V3 work out of the box with these sorts of themes?

Some notes from my own past experience, https://www.google.com/design/spec/components/text-fields.html#text-fields-labels - select 2 looks hideous out of the box and the div layout breaks form field label behaviour.
http://fezvrasta.github.io/bootstrap-material-design/#input - only supports selectize and once again looks hideous.

**Comment by kliver** [comment]
Using autocomplete for input type = text was really great!

I loved Global Navigation (http://django-autocomplete-light.readthedocs.org/en/2.3.1/navigation.html) and some other uses.

V3 is great but as you said, we loss some great customization features.

**Comment by jpic** [comment]
Well, it's a trade-off for everybody, currently v3 proposes an easier-to-maintain version by default. However, there is room for other widget scripts and form classes. What's important is that new users, new projects get the version that has less chances to fail, and that's best supported by Django.

I agree to have advanced features, just not by default anymore, because there's a lot of new users to support out there. The widgets as we have them in v2, are ridiculously simple to port, I've been talking about the new design we'd port them with in v3, but nobody answered.

If anybody wants to contribute, or couch host me a few days to sprint on a particular feature after the workday, then this will move in the right direction, but I fail to see how opening a ticket explaining why one version had better support for your use case than another could be of any help.

**Comment by jpic** [comment]
Not to mention that global navigation, is a feature of the autocomplete.js script by itstelf, and DAL itself doesn't provide *anything* special to use it. But yeah I'm all for all that, we just need more resource ;)

**Comment by techdragon** [comment]
Unfortunately I missed the opportunity to discuss the porting of the V2 widgets to V3 since I found and began using DAL during the brief window before V3 was completed and I didn't want to comment until I had used it more.

I'd offer to contribute couch or code, but the situation that prompted me to open this ticket is currently depriving me of any time to spare on other projects, and I doubt a Couch in Western Australia is convenient for sprinting on this feature after the workday. 

As for how I thought opening an issue like this would help instead of a more terse feature request.
Explaining how the new Select2 widget was inferior for the use cases in question, was an attempt to not insult all the hard work that had just gone into releasing V3 while I was asking for something that was removed to be returned. I wasn't sure if anyone else felt the same way about the V2 widgets, if there was ways to deal with the problem using Select2, if anything that changed during the extensive rewrite would make the old widgets unusable. I had a lot of unknowns to deal with so I tried to write an issue that might answer as many of them as possible.

**Comment by jpic** [comment]
>  I'd offer to contribute couch or code, but the situation that prompted
me to open this ticket is currently depriving me of any time to spare on
other projects, and I doubt a Couch in Western Australia is convenient for
sprinting on this feature after the workday.

Why not maybe just use v2 until this feature gets in v3 ?

Does this work to get the latest v2 ?

pip install -U 'django-autocomplete-light<3.0'

**Comment by techdragon** [comment]
@jpic Thats my current solution. 

Also, I think this would be better off closed, replaced with an issue for "Porting the V2 widgets to V3" since from your comments about them they don't sound too hard to port. Would you prefer a new issue and PR together in the future when I can find time? or open the issue now to let anyone else who might want to contribute before then know this would be desirable?

**Comment by jpic** [comment]
First things first, design !

#579 is a PoC of the ModelFormMetaclass that allows form field classes to
register themselves for auto-discovery, to override the default form fields
generated by the modelform. So that would be IMHO a sane way to customize
how we get ModelForm to generate form fields. Could you review this and
tell me if that would be the way to go, if so maybe you could help complete
this.

#589 is a PoC for the replacement of the "Autocomplete" classes, which sort
of "break" Django's separation of url / view / field / widget, but allows
to re-use methods in the different cases and have better DRY configuration.
Here we've got a metaclass to have views auto-register themselves in
urls.urlpatterns could you perhaps review this and let me know if that
looks like the right way to go and perhaps contribute to complete this.
Also, these SmartViews should be able to register themselves to the
ModelFormMetaclass and be able to spit out form fields themselves, but
that's after #579 is completed.

Another issue, perhaps easier, is that currently we have pretty rough basic
views (
https://github.com/yourlabs/django-autocomplete-light/issues/571#issuecomment-179957729
) so what would be great would be to have configurable views (ie. with
search_fields and this kind of things).

Thanks a heap for getting involved, really appreciatied !

**Comment by litchfield** [comment]
For selectize support, see #716

**Comment by jpic** [comment]
We've got a poc started in #749 , please take discussion there.
